src/encoder.rs

```diff
@@ -807,140 +808,580 @@ impl<W: Write> Drop for Writer<W> {
+        //               is 64 bits.
+        const CAP: usize = std::u32::MAX as usize >> 1;
+        let curr_chunk;
+        if writer.sep_def_img || writer.info.frame_control.is_none() || writer.written == 0 {
+            curr_chunk = chunk::IDAT;
+        } else {
+            curr_chunk = chunk::fdAT;
+        }
         ChunkWriter {
             writer,
-            buffer: vec![0; buf_len],
+            buffer: vec![0; CAP.min(buf_len)],
             index: 0,
+            curr_chunk,
         }
     }
-}
 
-impl<'a, W: Write> AsMut<Writer<W>> for ChunkOutput<'a, W> {
-    fn as_mut(&mut self) -> &mut Writer<W> {
-        match self {
-            ChunkOutput::Borrowed(writer) => writer,
-            ChunkOutput::Owned(writer) => writer,
+    /// Returns the size of each scanline for the next frame
+    /// paired with the size of the whole frame
+    ///
+    /// This is used by the `StreamWriter` to know when the scanline ends
+    /// so it can filter compress it and also to know when to start
+    /// the next one
+    fn next_frame_info(&self) -> (usize, usize) {
+        let wrt = self.writer.deref();
+
+        let width: usize;
+        let height: usize;
+        if let Some(fctl) = wrt.info.frame_control {
+            width = fctl.width as usize;
+            height = fctl.height as usize;
+        } else {
+            width = wrt.info.width as usize;
+            height = wrt.info.height as usize;
         }
+
+        let in_len = wrt.info.raw_row_length_from_width(width as u32) - 1;
+        let data_size = in_len * height;
+
+        (in_len, data_size)
     }
-}
 
-impl<W: Write> Write for ChunkWriter<'_, W> {
-    fn write(&mut self, mut buf: &[u8]) -> io::Result<usize> {
-        let written = buf.read(&mut self.buffer[self.index..])?;
-        self.index += written;
+    /// NOTE: this bypasses the internal buffer so the flush method should be called before this
+    ///       in the case there is some data left in the buffer when this is called, it will panic
+    fn write_header(&mut self) -> Result<()> {
+        if self.index != 0 {
+            // return Err(todo!())
+            panic!("Called when not flushed") // TODO: add error code
+        }
+        let wrt = self.writer.deref_mut();
 
-        if self.index >= self.buffer.len() {
-            self.writer
-                .as_mut()
-                .write_chunk(chunk::IDAT, &self.buffer)?;
-            self.index = 0;
+        if wrt.sep_def_img || wrt.info.frame_control.is_none() {
+            self.curr_chunk = chunk::IDAT;
+        } else {
+            if wrt.written == 0 {
+                self.curr_chunk = chunk::IDAT;
+            } else {
+                self.curr_chunk = chunk::fdAT;
+            }
+            if let Some(ref mut fctl) = wrt.info.frame_control {
+                fctl.encode(&mut wrt.w)?;
+                fctl.sequence_number += 1;
+            } else {
+                unreachable!()
+            }
```

Comment:
This would be clarer by matching on `wrt.info.frame_control`.
```suggestion
        match wrt.info.frame_control {
            _ if wrt.sep_def_mg => self.curr_chunk = chunk::IDAT,
            None => self.curr_chunk = chunk::IDAT,
            Some(ref mut fctl) => {
                // …
            }
        }
```